pkg/ebpf/ingress_node_firewall_loader.go

```diff
@@ -55,43 +59,295 @@ func NewIngNodeFwController() (*IngNodeFwController, error) {
+//      are updated.
+// vi)  Generate ingress node firewall events.
+// In the context of this method, stale keys are keys that figure inside the eBPF map but that are not generated
+// during step ii) from the provided ingressRules slice.
+func (infc *IngNodeFwController) IngressNodeFwRulesLoader(
+	ifaceIngressRules map[string][]v1alpha1.IngressNodeFirewallRules) error {
+	// Get eBPF objs to create/update eBPF maps and get map info.
+	info, err := infc.objs.BpfMaps.IngressNodeFirewallTableMap.Info()
+	if err != nil {
+		return fmt.Errorf("cannot get map info: %v", err)
+	}
+	klog.Infof("Ingress node firewall map Info: %+v with FD %s", info, infc.objs.BpfMaps.IngressNodeFirewallTableMap.String())
+
+	// Convert IngressNodeFirewallRules into data that can be written to the BPF map.
+	// Build a map of valid ebpfKeys pointing to the ebpfRules that should be associated to them.
+	ebpfKeyToRules := make(map[BpfLpmIpKeySt]BpfRulesValSt)
+	for interfaceName, ingressRules := range ifaceIngressRules {
+		// Look up the network interface by name.
+		iface, err := net.InterfaceByName(interfaceName)
+		if err != nil {
+			return fmt.Errorf("lookup error for network iface %q: %s", interfaceName, err)
+		}
+		ifID := uint32(iface.Index)
+
+		// Convert each provided ingressRule into a mapping of potentially multiple keys (one for each CIDR)
+		// pointing to a flattened rule that can be written to the BPF map.
+		for _, rule := range ingressRules {
+			if ebpfKeys, ebpfRules, err := infc.makeIngressFwRulesMap(rule, ifID); err == nil {
+				for _, ebpfKey := range ebpfKeys {
+					ebpfKeyToRules[ebpfKey] = ebpfRules
+				}
+			} else {
+				return fmt.Errorf("failed to create map firewall rules: %v on if %d", err, ifID)
+			}
+		}
+	}
+
+	// Build a slice of desired keys - it's easier to iterate over this slice later.
+	var desiredKeys []BpfLpmIpKeySt
+	for desiredKey := range ebpfKeyToRules {
+		desiredKeys = append(desiredKeys, desiredKey)
+	}
+
+	// Get stale keys (= keys inside the eBPF map but not inside the currently desired ruleset). Those keys
+	// will be dropped in the next step.
+	staleKeys, err := infc.getStaleKeys(desiredKeys)
 	if err != nil {
-		return fmt.Errorf("Cannot get map info: %v on if %d", err, ifId)
+		return err
+	}
+
+	// Purge all stale keys from the eBPF map.
+	if err := infc.purgeKeys(staleKeys); err != nil {
+		klog.Infof("Purge keys operation encountered issues, err: %q", err)
 	}
-	klog.Infof("Ingress node firewall map Info: %+v with FD %s", info, objs.BpfMaps.IngressNodeFirewallTableMap.String())
 
-	if err := infc.makeIngressFwRulesMap(ingFireWallConfig, isDelete, ifId); err != nil {
-		return fmt.Errorf("Failed to create map firewall rules: %v on if %d", err, ifId)
+	// Add/update all keys. This is an idempotent action and non-existing keys are added whereas existing keys
+	// are updated.
+	if err := infc.addOrUpdateRules(ebpfKeyToRules); err != nil {
+		return err
 	}
 
+	// Generate ingress node few events.
```

Comment:
can u move this to NewController we just need to start evens loop once else just put FIXME and I will move it in future PRs

Reply:
I moved it